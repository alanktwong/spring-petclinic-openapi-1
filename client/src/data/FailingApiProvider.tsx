import * as React from 'react';
import { FailingApi } from 'petclinic-api';
import { Subtract } from 'utility-types';

export interface WithFailingApiProps {
  failingApi: FailingApi;
}

export const FailingApiContext = React.createContext(new FailingApi());

export const withFailingApi = <P extends WithFailingApiProps>(
  Component: React.ComponentType<P>
) =>
  class WithFailingApi extends React.Component<
    Subtract<P, WithFailingApiProps>
  > {
    static contextType = FailingApiContext;
    context!: React.ContextType<typeof FailingApiContext>;

    render() {
      return <Component failingApi={this.context} {...(this.props as P)} />;
    }
  };
