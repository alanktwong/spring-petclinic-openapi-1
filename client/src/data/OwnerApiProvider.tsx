import * as React from 'react';
import { OwnerApi } from 'petclinic-api';
import { Subtract } from 'utility-types';

export interface WithOwnerApiProps {
  ownerApi: OwnerApi;
}

export const OwnerApiContext = React.createContext(new OwnerApi());

export const withOwnerApi = <P extends WithOwnerApiProps>(
  Component: React.ComponentType<P>
) =>
  class WithOwnerApi extends React.Component<Subtract<P, WithOwnerApiProps>> {
    static contextType = OwnerApiContext;
    context!: React.ContextType<typeof OwnerApiContext>;

    render() {
      return <Component ownerApi={this.context} {...(this.props as P)} />;
    }
  };
