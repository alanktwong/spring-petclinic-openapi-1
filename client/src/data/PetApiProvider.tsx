import * as React from 'react';
import { PetApi } from 'petclinic-api';
import { Subtract } from 'utility-types';

export interface WithPetApiProps {
  petApi: PetApi;
}

export const PetApiContext = React.createContext(new PetApi());

export const withPetApi = <P extends WithPetApiProps>(
  Component: React.ComponentType<P>
) =>
  class WithPetApi extends React.Component<Subtract<P, WithPetApiProps>> {
    static contextType = PetApiContext;
    context!: React.ContextType<typeof PetApiContext>;

    render() {
      return <Component petApi={this.context} {...(this.props as P)} />;
    }
  };
