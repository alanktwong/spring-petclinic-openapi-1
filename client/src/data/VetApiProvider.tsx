import * as React from 'react';
import { VetApi } from 'petclinic-api';
import { Subtract } from 'utility-types';

export interface WithVetApiProps {
  vetApi: VetApi;
}

export const VetApiContext = React.createContext(new VetApi());

export const withVetApi = <P extends WithVetApiProps>(
  Component: React.ComponentType<P>
) =>
  class WithVetApi extends React.Component<Subtract<P, WithVetApiProps>> {
    static contextType = VetApiContext;
    context!: React.ContextType<typeof VetApiContext>;

    render() {
      return <Component vetApi={this.context} {...(this.props as P)} />;
    }
  };
