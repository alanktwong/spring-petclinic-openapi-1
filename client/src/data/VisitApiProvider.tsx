import * as React from 'react';
import { VisitApi } from 'petclinic-api';
import { Subtract } from 'utility-types';

export interface WithVisitApiProps {
  visitApi: VisitApi;
}

export const VisitApiContext = React.createContext(new VisitApi());

export const withVisitApi = <P extends WithVisitApiProps>(
  Component: React.ComponentType<P>
) =>
  class WithVisitApi extends React.Component<Subtract<P, WithVisitApiProps>> {
    static contextType = VisitApiContext;
    context!: React.ContextType<typeof VisitApiContext>;

    render() {
      return <Component visitApi={this.context} {...(this.props as P)} />;
    }
  };
