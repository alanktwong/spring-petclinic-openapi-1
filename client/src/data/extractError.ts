import { IError } from '../types';

const extractError = (response): Promise<IError> => {
  if (response.json) {
    return response.json();
  } else {
    return Promise.resolve({ message: `${response}` });
  }
};

export default extractError;
