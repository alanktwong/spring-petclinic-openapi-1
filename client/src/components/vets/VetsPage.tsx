import * as React from 'react';

import { RouteComponentProps, withRouter } from 'react-router-dom';

import { Vet } from 'petclinic-api';
import { withVetApi, WithVetApiProps } from '../../data/VetApiProvider';

import { IError } from '../../types';
import PageErrorMessage from '../PageErrorMessage';
import extractError from '../../data/extractError';

import Loading from '../Loading';

interface IVetsPageState {
  error?: IError;
  vets?: Vet[];
}

interface IVetsPageProps extends RouteComponentProps, WithVetApiProps {}

class VetsPage extends React.Component<IVetsPageProps, IVetsPageState> {
  async componentDidMount() {
    try {
      const vets = await this.props.vetApi.listVets({});

      console.log('vets', vets);
      this.setState({ vets });
    } catch (response) {
      const error = await extractError(response);
      this.setState({ error });
    }
  }

  render() {
    const { error, vets } = this.state || {};

    if (error) {
      return <PageErrorMessage error={error} />;
    }

    if (!vets) {
      return <Loading />;
    }

    return (
      <div>
        <h2>Veterinarians</h2>
        <table className="table table-striped">
          <thead>
            <tr>
              <th>Name</th>
              <th>Specialties</th>
            </tr>
          </thead>
          <tbody>
            {vets.map(vet => (
              <tr key={vet.id}>
                <td>
                  {vet.firstName} {vet.lastName}
                </td>
                <td>
                  {vet.specialties.length > 0
                    ? vet.specialties
                        .map(specialty => specialty.name)
                        .join(', ')
                    : 'none'}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
}

export default withVetApi(withRouter(VetsPage));
