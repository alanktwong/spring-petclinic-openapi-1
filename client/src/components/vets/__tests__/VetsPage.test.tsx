import React from 'react';
import renderer from 'react-test-renderer';
import VetsPage from '../VetsPage';
import waitForExpect from 'wait-for-expect';
import { VetApiContext } from '../../../data/VetApiProvider';
import { VetApi } from 'petclinic-api';
import { StaticRouter } from 'react-router-dom';

it('renders correctly', async () => {
  const mock = new VetApi();
  mock.listVets = async () => [
    {
      id: 1,
      firstName: 'firstName1',
      lastName: 'lastName1',
      specialties: [
        {
          id: 2,
          name: 'specialtyName2',
        },
      ],
    },
  ];

  const component = renderer.create(
    <StaticRouter>
      <VetApiContext.Provider value={mock}>
        <VetsPage />
      </VetApiContext.Provider>
    </StaticRouter>
  );

  const page = component.root.find(node => node.type.name === 'VetsPage');

  await waitForExpect(() => {
    expect(page.instance.state.vets[0].id).toBe(1);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
