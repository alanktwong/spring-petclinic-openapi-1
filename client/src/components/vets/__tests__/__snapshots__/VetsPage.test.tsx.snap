// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h2>
    Veterinarians
  </h2>
  <table
    className="table table-striped"
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Specialties
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          firstName1
           
          lastName1
        </td>
        <td>
          specialtyName2
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
